### ISSUE TITLE
Increase sharing for in-memory headers

### ISSUE THREAD

**Issue by mrBliss** [input-output-hk/ouroboros-network#505]
1. The ChainSync Client receives headers from the network layer. Currently, for each candidate, all headers are kept in memory. Even after #465 a large number of headers will be kept in memory.
2. The ChainDB will keep the headers of blocks in memory to perform chain selection. These headers are extracted from the blocks, which are read from disk.

Unfortunately,  each header is kept in memory *twice*, because they are deserialised twice, so they won't be pointers to the same chunk of memory.  If each header is 1kB and there are ~2k headers for each node in memory, we have 2x ~2MB of headers *per node*.

By sharing these headers between the ChainSync Client and the ChainDB, we can halve this. For example, by "interning" headers using a header table, a map from point to header.

**Comment by edsko** [comment]
The Chain DB keeps the _current_ chain in memory (as headers). In the typical case, when we have `n` upstream nodes, al of which are following the same chain, introducing this cache would reduce memory usage from `(1 + n) * 2 MB` to just a constant 2 MB.

**Comment by edsko** [comment]
We don't want to optimize for the best case. Closing.

**Comment by edsko** [comment]
For leaf nodes only -- the above objection is appropriate for servers only. But low priority, for sure.
